docs/network-drivers.md

```diff
@@ -213,6 +215,57 @@ The body of the api call request will look like:
 The brocade driver accepts interface names the same way they would be accepted
 in the console of the switch, ex. ``101/0/10``.
 
+### Juniper QFX5200 driver
+
+#### Switch preparation
+
+1. The driver relies on netconf protocol to manage the switch. 
+It is important to configure switch to enable netconf connection.
+With administration privileges, running the following command in 
+configuration mode should set it up for netconf based connections. 
+
+```
+set system services netconf ssh
+```
+
+2. Driver relies on the PyEZ library, a python wrapper that for netconf 
+using `ncc client`. Make sure it is installed before using the driver. 
+
+3. VLANS that HIL will manage should be preconfigured on switch.
+For each VLAN that HIL will manage, create it on the switch using following commands.
+To create a vlan with name `hil-500` having vlan-id 500, do the following,
+in configuration mode run:
+
+```
+set vlans hil-500 vlan-id 500
+``` 
+
+4. Configure the initial state of the interface (ports) configuration manually.
+Then copy the configuration as a jinja2 template and store it at
+`hil/ext/switches/junos/jinja_templates`. This will by used by `revert_port` 
+to reset the ports to its initial conditions when required. 
```

Comment:
This is not a reasonable interface. There are a few problems:

1. In general, users should not be storing things inside the HIL source tree. Remember, HIL may actually be installed somewhere, and not using the source tree directly, and the equivalent thing would be asking them to shove a file somewhere in `/usr/lib/python2.7/site-packages`, which is... horrible.
2. Sysadmins should not be asked to mess with templating languages, especially as part of routine set-up.

Can you point to some documentation explaining how all this works? There has to be a simpler way to manage this.

Reply:
This is part of how Juniper uses the pyEZ library.  It is quite simple to use. I have provided the comments in the templates itself. I am assuming that the one who is setting up the initial switch configuration has a good idea of what a default port state should look like. They have to mimic that state in the jinja template. It is not complex at all.  A simpler way is I try to pass the template as a string or json or dictionary. In any case the sysadmin who will have to make some changes to the driver to decide what a default port state should look like. I thought this was cleaner and readable way of doing things. 
here is the link: https://www.juniper.net/documentation/en_US/junos-pyez/topics/task/program/junos-pyez-program-configuration-data-loading.htmlhttps://www.juniper.net/documentation/en_US/junos-pyez/topics/task/program/junos-pyez-program-configuration-data-loading.html

Reply:
Ok, I'll read up on the documentation and weigh in again. I am regardless not open to storing the template (or any user-supplied data) in the source tree, but that can maybe go in the database or something, if it's the right way to go. I might be amenable if this is really the way juniper expects you to manage the switch normally.

Reply:
For port revert there are 2 files and those contain some string and comments. Why can't we just put that string in the driver (the python file) itself?

Reply:
I had to learn using `jinja` templates because pyez library relies on it, although there are other ways of getting the same results. It is simpler to make changes using templates which otherwise would require multiple `set` and `delete` statements to be executed on the switch. @naved001 I kept it in a separate file so as to hide the complexity from the user. The user does not need to go through the driver code to figure out how to set initial port of the driver. @zenhack Juniper has chosen `jinja templating` because, it can do lot of complex things with it. I provide a simpler template but complex templating including`for loops` and other constructs can be used to create a desired state. That is another reason I decided to make this its own file. That way it keeps things readable in their own files.

Reply:
We should let the admin set the path for the templates in the config file (under a section for the juniper siwtch `[hil.ext.switch.juniper]`) and the driver then reads the path from there and loads them jinja templates. @henn thought this was a reasonable solution. What do you think @zenhack ?

Reply:
Having picked through the documentation a bit, it's still not clear to me why we need to have the user supply a template at all -- the operations we perform are pretty cookie-cutter, I would think we would just be able to write a template (or a few of them) ahead of time, and pass data to it at run-time.